chips/stm32f4xx/src/chip_specific.rs

```diff
@@ -0,0 +1,166 @@
+    const __APB1_FREQUENCY_LIMIT_MHZ: usize = 42;
+
+    /// Maximum allowed APB1 frequency in MHz
+    pub const APB1_FREQUENCY_LIMIT_MHZ: usize = __APB1_FREQUENCY_LIMIT_MHZ;
+
+    /// Maximum allowed APB2 frequency in MHz
+    // APB2 frequency limit is twice the APB1 frequency limit
+    pub const APB2_FREQUENCY_LIMIT_MHZ: usize = APB1_FREQUENCY_LIMIT_MHZ << 1;
+
+    #[cfg(any(feature = "stm32f412"))]
+    const __SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = 100;
+    #[cfg(any(feature = "stm32f429", feature = "stm32f446"))]
+    #[cfg(not(feature = "cargo-clippy"))]
+    const __SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = 168;
+    #[cfg(any(feature = "stm32f401"))]
+    #[cfg(not(feature = "cargo-clippy"))]
+    const __SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = 84;
+
+    /// Maximum allowed system clock frequency in MHz
+    pub const SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = __SYS_CLOCK_FREQUENCY_LIMIT_MHZ;
+}
+
+/// Chip-specific flash code
+pub mod flash_specific {
+    #[cfg(any(
+        feature = "stm32f401",
+        feature = "stm32f412",
+        feature = "stm32f429",
+        feature = "stm32f446"
+    ))]
+    #[derive(Copy, Clone, PartialEq, Debug)]
+    /// Enum representing all the possible values for the flash latency
+    pub(crate) enum FlashLatency {
+        /// 0 wait cycles
+        Latency0,
+        /// 1 wait cycle
+        Latency1,
+        /// 2 wait cycles
+        Latency2,
+        /// 3 wait cycles
+        Latency3,
+        /// 4 wait cycles
+        Latency4,
+        /// 5 wait cycles
+        Latency5,
+        /// 6 wait cycles
+        Latency6,
+        /// 7 wait cycles
+        Latency7,
+        /// 8 wait cycles
+        Latency8,
+        /// 9 wait cycles
+        Latency9,
+        /// 10 wait cycles
+        Latency10,
+        /// 11 wait cycles
+        Latency11,
+        /// 12 wait cycles
+        Latency12,
+        /// 13 wait cycles
+        Latency13,
+        /// 14 wait cycles
+        Latency14,
+        /// 15 wait cycles
+        Latency15,
+    }
+
+    // The number of wait cycles depends on two factors: system clock frequency and the supply
+    // voltage. Currently, this method assumes 2.7-3.6V voltage supply (default value).
+    // TODO: Take into the account the power supply
+    //
+    // The number of wait states varies from chip to chip.
+    pub(crate) fn get_number_wait_cycles_based_on_frequency(frequency_mhz: usize) -> FlashLatency {
+        // feature = "stm32f401"
+        // feature = "stm32f429"
+        // feature = "stm32f446"
+        #[cfg(not(feature = "stm32f412"))]
+        {
+            match frequency_mhz {
+                0..=30 => FlashLatency::Latency0,
+                31..=60 => FlashLatency::Latency1,
+                61..=90 => FlashLatency::Latency2,
+                91..=120 => FlashLatency::Latency3,
+                121..=150 => FlashLatency::Latency4,
+                _ => FlashLatency::Latency5,
+            }
+        }
```

Comment:
For example, this is the default setting for `get_number_wait_cycles_based_on_frequency`, correct? So if a user sets no features, they get these values, or if a user tries a new board with a different stm32f4 variant and sets a different feature they get these values? Are these the values for every stm32f4 variant (now and into the future) not called "stm32f412"? If the datasheet says so then that is good enough for me. If STM changes the datasheet in the future we can't realistically plan for that.

The purpose of our crate-based approach is we don't have these ambiguities. Everything in the stm32f4xx crate should apply to all stm32f4 chips. Everything in the stm32f446re crate only needs to apply to that specific variant. If there is anything in the stm32f4xx crate that could be used on the wrong variant that is a bug in my opinion.

Reply:
I agree with you. Check the second link I sent you. I think it solves this issue. The only place where features are not explicitly stated is `PLL_MIN_FREQ_MHZ`, since only STM32F401 supports 13MHz as the minimum PLL frequency. All the other chips in the family support 24MHz.

Reply:
I think the approach in the second link looks pretty reasonable. I'm not sure I have a strong opinion between `#[cfg` and `cfg!(`. The biggest negative of the latter is that it can 'hide' in the middle of lines (`if (...blah && cfg!(..`), but when contained to the `chip_specific` file already, where it's more sane to be on the lookout for `cfg` stuff, I think it is a reasonable solution.

Reply:
I understand your concerns about `cfg!()` macros. However, used in `chip_specific.rs` and in simple if statements without any logical operators, they are easy to understand and easy to write, at least for me. Also, they don't require any dirty hacks to pass CI tests (or probably I missunderstand how to use `#[cfg]` attributes).